auction/program/src/processor.rs

```diff
@@ -0,0 +1,152 @@
+    OpenEdition,
+}
+
+/// Bidding Implementations.
+///
+/// English Auction: this stores only the current winning bids in the auction, pruning cancelled
+/// and lost bids over time.
+///
+/// Open Edition: All bids are accepted, cancellations return money to the bidder and always
+/// succeed.
+impl BidState {
+    fn new_english(n: usize) -> Self {
+        BidState::EnglishAuction {
+            bids: vec![],
+            max: n,
+        }
+    }
+
+    fn new_open_edition() -> Self {
+        BidState::OpenEdition
+    }
+
+    /// Push a new bid into the state, this succeeds only if the bid is larger than the current top
+    /// winner stored. Crappy list information to start with.
+    fn place_bid(&mut self, bid: Bid) -> Result<(), ProgramError> {
+        match self {
+            // In a capped auction, track the limited number of winners.
+            BidState::EnglishAuction { ref mut bids, max } => match bids.last() {
+                Some(top) if top.1 < bid.1 => {
+                    bids.retain(|b| b.0 != bid.0);
+                    bids.push(bid);
+                    if bids.len() > *max {
+                        bids.remove(0);
+                    }
+                    Ok(())
+                }
+                _ => Err(AuctionError::BidTooSmall.into()),
+            },
+
+            // In an open auction, bidding simply succeeds.
+            BidState::OpenEdition => Ok(()),
+        }
+    }
+
+    /// Cancels a bid, if the bid was a winning bid it is removed, if the bid is invalid the
+    /// function simple no-ops.
+    fn cancel_bid(&mut self, key: Pubkey) -> Result<(), ProgramError> {
+        match self {
+            BidState::EnglishAuction { ref mut bids, max } => {
+                bids.retain(|b| b.0 != key);
+                Ok(())
+            }
+
+            // In an open auction, cancelling simply succeeds. It's up to the manager of an auction
+            // to decide what to do with open edition bids.
+            BidState::OpenEdition => Ok(()),
+        }
+    }
+
+    /// Check if a pubkey is currently a winner.
+    fn is_winner(&self, key: Pubkey) -> bool {
+        match self {
+            // Presense in the winner list is enough to check win state.
+            BidState::EnglishAuction { ref bids, max } => bids.iter().any(|bid| bid.0 == key),
+            // There are no winners in an open edition, it is up to the auction manager to decide
+            // what to do with open edition bids.
+            BidState::OpenEdition => false,
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, BorshSerialize, BorshDeserialize, PartialEq)]
+pub enum WinnerLimit {
+    Unlimited,
+    Capped(usize),
+}
+
+/// Models a set of metadata for a bidder, meant to be stored in a PDA. This allows looking up
+/// information about a bidder regardless of if they have won, lost or cancelled.
+struct BidderMetadata {
+    /// Tracks the last time this user bid.
+    last_bid: UnixTimestamp,
```

Comment:
struct LastBid
- Add public key of user bidder.
- Bid amount is a better use of this variable.
- last_bid_timestamp: UnixTimestamp

Reply:
Store cancel in both meta and bistate.